### PR TITLE
[FEA] Add support for different eval dataloaders for valid and test phases.

### PR DESCRIPTION
**Pull request by ChenglongMa**
Pull request for https://github.com/RUCAIBox/RecBole/issues/1640

# Main changelogs

1. Moved `self.is_sequential` to `AbstractDataLoader` (will be used in `case study`);
2. Fixed `tensors used as indices` error when `dtype` is not `long` (the default `dtype` in `torch.tensor` is `torch.int32`);
3. Added unit tests and fix some minor errors;
4. Added different `mode` for valid and test phases in `eval_args` config;
5. Updated docs.

### CODE REVIEW

**Comment by Sherry-XLL** [comment]
Hello @ChenglongMa! 

Thanks for your attention and contributions to RecBole! Since there are many changes involved, we will spend some time to review the rationality of the code structure and the convenience of the change. The update may be merged into the main code in the later time or next version. 

Thank you again for your support to our team!

File: docs/source/get_started/distributed_training.rst [hunk 1]
```diff
@@ -269,7 +269,7 @@ You can create a `yaml` file called `test.yaml` and set all the config as follow
         group_by: user
         order: RO
         split: {'RS': [0.8,0.1,0.1]}
-        mode: full
+        mode: {"valid": "full", "test": "full"}
```
Comment (Sherry-XLL) [inline docs/source/get_started/distributed_training.rst:272]: Hello, @ChenglongMa. What wonderful code! Thanks very much for your attention to RecBole!

As for the `mode` in `eval_args`, maybe we can keep the original setting, but the interface of the dictionary is extended.  In other words, we still allow users to input `full` or `uni100` as the mode. In this case, the evaluation and test modes are the same. At the same time, we also provide a dictionary interface. If the user inputs the form of the dictionary, we judge whether there are `valid` and `test` key values, and set the corresponding evaluation mode. That is to say, only when different modes are needed for evaluation and testing, the dictionary form is needed to facilitate compatibility with previous configurations.

We would appreciate it if it would be convenient for you to make further modifications. Thank you again for your contributions to RecBole!

File: recbole/config/configurator.py [hunk 2]
```diff
@@ -440,21 +442,29 @@ def _set_default_parameters(self):
...
+        # backward compatible
+        if isinstance(mode, str):
+            default_eval_args['mode'] = {'valid': mode, 'test': mode}
+
+        # in case there is only one key in `mode`, e.g., mode: {'valid': 'uni100'} or mode: {'test': 'full'}
+        if isinstance(mode, dict):
+            default_mode = mode.get('valid', mode.get('test', 'full'))
+            default_eval_args['mode'] = {'valid': mode.get('valid', default_mode), 'test': mode.get('test', default_mode)}
+
+        self.final_config_dict["eval_args"] = default_eval_args
```
Comment (ChenglongMa) [inline recbole/config/configurator.py:464]: Hi @Sherry-XLL, thank you for your kind words. I would be very happy if I can make a contribution to such a brilliant library.

Regarding your question, these pieces of code can explain more. Here it supports three scenarios:

1.  The original settings, e.g., `mode: full`. It will be converted into `"mode": {"valid": "full", "test": "full"}` in the code.
2. There is only **ONE** key in the settings, e.g., `mode: {valid: full}` or `mode: {test: full}`. It will be extended to TWO keys with the same `mode`: `"mode": {"valid": "full", "test": "full"}`;
3. There are **TWO** keys in the settings, which is the expected version.

One more thing, we here assume users can only use `valid` or `test` as the key in the settings, otherwise, it will raise a `ValueError`.

Thanks!

Comment (Sherry-XLL) [reply recbole/config/configurator.py:464]: Hello, @ChenglongMa! 

Thank you for your timely update of the documentation! We are very happy to see your expanded function of the evaluation mode. As for the default `.rst` document (_e.g._, `distributed_training.rst`), we still prefer to keep it in the original `str` format. In other words, the default configuration in the document is still `full` or `uni100` instead of `{"valid": "full", "test": "full"}`, which is simple and clear in most cases. The dictionary configuration is only utilized when users want to use different modes for evaluation and test.

Thank you again for your great contributions to RecBole! If you can adjust the default configuration of the document at a convenient time, we are very happy to merge your code into the main branch.

Comment (ChenglongMa) [reply recbole/config/configurator.py:464]: Hi @Sherry-XLL, I've reverted the related docs and unit tests to the original versions. Thanks!

**Comment by Sherry-XLL** [comment]
Hello @ChenglongMa! We have merged your code into the master branch and thanks again for your great contributions to RecBole. Welcome to follow up the latest progress of our library and put forward more valuable suggestions. Looking forward to your next feedback!
